From c9391fafbdcd4b13e079fb312be42b2406ce37a1 Mon Sep 17 00:00:00 2001
From: Jing Zhao <jing9@apache.org>
Date: Wed, 25 Nov 2015 14:21:06 -0800
Subject: [PATCH 2292/2748] HDFS-9467. Fix data race accessing
 writeLockHeldTimeStamp in FSNamesystem.
 Contributed by Mingliang Liu.

(cherry picked from commit e556c35b0596700f9ec9d0a51cf5027259d531b5)
(cherry picked from commit 288cf8437b7e03f071e95eb05e83a26e58fff26b)
(cherry picked from commit 8aa18534d9d1bee03e3a5b8db15d2b2034e3adc6)

Change-Id: Id7e715e1d75c3dd78b40f916e4ffe81a0e622a8c
---
 .../hadoop/hdfs/server/namenode/FSNamesystem.java  |   11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
index 3e4d994..e686bc6 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
@@ -1636,14 +1636,13 @@ public void writeLockInterruptibly() throws InterruptedException {
   public void writeUnlock() {
     final boolean needReport = fsLock.getWriteHoldCount() == 1 &&
         fsLock.isWriteLockedByCurrentThread();
+    final long writeLockInterval = monotonicNow() - writeLockHeldTimeStamp;
+
     this.fsLock.writeLock().unlock();
 
-    if (needReport) {
-      long writeLockInterval = monotonicNow() - writeLockHeldTimeStamp;
-      if (writeLockInterval >= WRITELOCK_REPORTING_THRESHOLD) {
-        LOG.info("FSNamesystem write lock held for " + writeLockInterval +
-            " ms via\n" + StringUtils.getStackTrace(Thread.currentThread()));
-      }
+    if (needReport && writeLockInterval >= WRITELOCK_REPORTING_THRESHOLD) {
+      LOG.info("FSNamesystem write lock held for " + writeLockInterval +
+          " ms via\n" + StringUtils.getStackTrace(Thread.currentThread()));
     }
   }
   @Override
-- 
1.7.9.5

